pytorch_lightning/core/optimizer.py

```diff
@@ -186,16 +188,35 @@ def dis_closure():
             make_optimizer_step = not self._should_accumulate
 
         trainer = self._trainer
+        model = trainer.get_model()
         optimizer = self._optimizer
 
         if make_optimizer_step:
+
+            if self._optimizer_idx is None:
+                for opt_idx, opt in enumerate(trainer.optimizers):
+                    if self == opt:
+                        self._optimizer_idx = opt_idx
+                        break
+
+            ddp_plugin = trainer.accelerator_backend.ddp_plugin
+            if ddp_plugin is not None and ddp_plugin.using_rpc:
+                if not self._in_foreach_worker and trainer.is_master:
```

Comment:
If we put the variable into the pipe_rpc_plugin, maybe we can strip out these if statements etc to clean things up here.

something like:

`if self.ddp_plugin.signal_workers_optimizer_step(trainer.is_master)`:

and the setter becomes `self.ddp_plugin._in_foreach_worker=True` etc

Reply:
Done !